### PR TITLE
Add checkpoint support

### LINKED ISSUE #92: Add checkpoint support
### Summary

Introduce checkpoint-based navigation methods that allow the router to return to a previously marked location in the `historyStack` by removing all views above it.

Instead of targeting a route path directly, routes can register **checkpoint identifiers**. The router can later unwind the stack back to the most recent matching checkpoint.

This enables clean handling of temporary navigation flows (such as
authentication, onboarding, or modal workflows) while preserving
existing route instances.

------------------------------------------------------------------------

### Proposed API

``` brightscript
sgRouter.setCheckPoint(identifier = invalid as Dynamic)

sgRouter.popToCheckPoint(identifier = invalid as Dynamic)
```

------------------------------------------------------------------------

#### `setCheckPoint(identifier)`

Marks the **currently active route** as a checkpoint.

-   If an identifier is provided, it is registered as a checkpoint
    identifier for the current route.
-   If no identifier is provided:
    -   The router calls `sgRouter.getActiveRoute()`
    -   The active route's `path` property is used as the checkpoint
        identifier.

A route may register **multiple checkpoint identifiers**.

Example:

``` brightscript
sgRouter.setCheckPoint("authStart")
sgRouter.setCheckPoint("returnPoint")
```

Both identifiers will be associated with the same route instance.

Calling `setCheckPoint()` multiple times with different identifiers does
**not overwrite** previous checkpoints.

Duplicate identifiers on the same route should be ignored.

------------------------------------------------------------------------

#### `popToCheckPoint(identifier)`

Unwinds the router's `historyStack` back to a previously registered checkpoint.

If an identifier is provided:

1.  The router searches the stack **from top to bottom**.
2.  The most recent route containing that checkpoint identifier is
    located.
3.  All views above that route are closed.
4.  The matched route becomes the active view.

If **no identifier is provided**:

1.  The router searches for the **most recent route that registered any
    checkpoint**.
2.  The stack unwinds to that route.

If no matching checkpoint exists, the operation is cancelled and no
navigation occurs.

------------------------------------------------------------------------

### Stack Behavior

When unwinding the stack:

1.  All views above the matched checkpoint route are removed.
2.  Each popped view runs its lifecycle hooks:
    -   `beforeViewClose`
    -   `onViewClose`
3.  The checkpoint route remains in the stack and becomes the active
    route.

------------------------------------------------------------------------

### KeepAlive Behavior

Routes removed during `popToCheckPoint()` must continue to respect
existing `keepAlive` rules.

If a route being removed has `keepAlive` enabled, it should follow the
router's existing keepAlive retention behavior exactly as it does for
other stack pops.

Checkpoint navigation **does not alter lifecycle or disposal
semantics**, only how the pop target is selected.

------------------------------------------------------------------------

### Checkpoint Identifier Rules

-   A route may hold **multiple checkpoint identifiers**
-   Identifiers must be unique per route
-   Calling `setCheckPoint()` without an identifier uses the active
    route's `path`
-   Identifier matching is **exact**

Example route metadata conceptually:

    Route: /movies/123

    checkpointIds:
        "/movies/123"
        "authReturn"
        "resumePlayback"

------------------------------------------------------------------------

#### Example 1: Implicit checkpoint using route path

Stack:

    /home
    /movies
    /movies/123

Calling:

``` brightscript
sgRouter.setCheckPoint()
```

Registers:

    "/movies/123"

as a checkpoint identifier.

------------------------------------------------------------------------

#### Example 2: Named checkpoint

Stack:

    /home
    /cart
    /checkout
    /payment

Calling:

``` brightscript
sgRouter.setCheckPoint("checkoutStart")
```

Later:

``` brightscript
sgRouter.popToCheckPoint("checkoutStart")
```

Results in:

    /home
    /cart
    /checkout

------------------------------------------------------------------------

#### Example 3: Pop to most recent checkpoint

Stack:

    /home
    /search
    /product
    /auth
    /profile

Checkpoints:

    /search -> "searchEntry"
    /auth   -> "authStart"

Calling:

``` brightscript
sgRouter.popToCheckPoint()
```

Results in:

    /home
    /search
    /product
    /auth

Because `/auth` registered the most recent checkpoint.

------------------------------------------------------------------------

### Use Case: Authentication Flow

A common scenario is authentication during playback.

1.  User deep links to `/movies/123`
2.  Movie details screen loads
3.  Route registers checkpoint

``` brightscript
sgRouter.setCheckPoint()
```

Possible stack:

    /movies/123
    /signin
    /verify-email
    /profiles
    /profile-pin

Once authentication completes:

``` brightscript
sgRouter.popToCheckPoint()
```

The router closes all authentication screens and restores the original
movie details view.

------------------------------------------------------------------------

### Benefits

-   Preserves the original screen instance
-   Avoids unnecessary view recreation
-   Supports complex navigation flows
-   Enables reusable unwind points in the stack
-   Simplifies interruption flows (authentication, onboarding, modals)

### PR DESCRIPTION
**Pull request by iObject**
## Branch: `92-add-popto-support`

This branch adds **checkpoint-based stack unwinding** to the router — a way to mark a point in the navigation history and later jump back to it in one call, without repeatedly calling `goBack()`.

### New public API

**`sgRouter.setCheckpoint(identifier)`** — stamps the current active stack entry with a named checkpoint. The identifier defaults to the active route's path if omitted. Multiple checkpoints can be registered on the same entry; duplicates are silently ignored.

**`sgRouter.popToCheckpoint(identifier)`** — unwinds the stack to the most recent entry carrying that checkpoint identifier. Views above the target are closed (keepAlive views are suspended rather than destroyed). Returns a promise that rejects if no matching checkpoint exists.

### Key implementation details

- Checkpoint metadata (`hasCheckpoint`, `checkpointIds`) is stored directly on history stack entries as an AA — no separate data structure.
- `setCheckpoint` returns a `Boolean` (true if newly stamped, false if duplicate or invalid).
- During a `popToCheckpoint`, `m.__router_processingGoBack = true` suppresses focus and history side-effects inside `showView` — the same flag used by `_goBack`, since both are backward navigations.
- History stack truncation happens **before** `showView` is called (same pattern as the inline `historyStack.pop()` used by `_goBack`), so `showView.finally` skips all history management for any backward navigation.

### Also on this branch

- `getActiveRoute()` — new public accessor (with tests).
- Test project updated with a signup flow (`SignupScreen`, `VerifyEmailScreen`, `SignupGuard`) demonstrating a real checkpoint use case: marking the start of a signup flow and popping back to it if the user needs to restart.
- 528 lines of new unit tests covering checkpoint set/pop behavior, keepAlive interaction, edge cases (no checkpoint found, empty stack, invalid identifier), and `getActiveRoute`.

### CODE REVIEW

File: src/components/Router.bs [hunk 1]
```diff
@@ -139,56 +156,168 @@ function _goBack(_ = Invalid as Dynamic) as Dynamic
...
     end sub)
 
     return true
 end function
 
+' Marks the current active stack entry with a checkpoint identifier.
+' If identifier is Invalid or omitted, the active route's path is used as the identifier.
+' Multiple identifiers may be registered on the same route; duplicates are ignored.
+' @param {Dynamic} identifier - A string checkpoint identifier, or Invalid to use the active route's path
+sub _setCheckPoint(identifier = Invalid as Dynamic)
```
Comment (TwitchBronBron) [inline src/components/Router.bs]: Probably use the `identifier = "__INVALID__"` pattern here (to differentiate between that and `invalid`/not specified)

File: src/components/Router.bs [hunk 2]
```diff
@@ -139,56 +156,168 @@ function _goBack(_ = Invalid as Dynamic) as Dynamic
...
+    if m.__router_historyStack.count() = 0 then return
```
Comment (TwitchBronBron) [inline src/components/Router.bs]: Should return `true` or `false` based on whether you were able to set a checkpoint.

File: src/components/Router.bs [hunk 3]
```diff
@@ -139,56 +156,168 @@ function _goBack(_ = Invalid as Dynamic) as Dynamic
...
+    lastIndex = m.__router_historyStack.count() - 1
+    entry = m.__router_historyStack[lastIndex]
+
+    ' Initialize checkpoint fields if not present on this entry
+    if entry.hasCheckpoint = Invalid then
+        entry.hasCheckpoint = false
+        entry.checkpointIds = {}
+    end if
+
+    ' Idempotent: skip if identifier already registered
```
Comment (TwitchBronBron) [inline src/components/Router.bs:200]: Could just check for .checkpointIds is not null and not empty AA (or just keep using bool, your call. both good)

File: src/components/Router.bs [hunk 4]
```diff
@@ -139,56 +156,168 @@ function _goBack(_ = Invalid as Dynamic) as Dynamic
...
+    if entry.checkpointIds[resolvedId] = true then return
+
+    entry.checkpointIds[resolvedId] = true
+    entry.hasCheckpoint = true
+end sub
```
Comment (TwitchBronBron) [inline src/components/Router.bs]: probably don't need this check? We'll accomplish it after this anyway...

File: src/components/Router.bs [hunk 5]
```diff
@@ -139,56 +156,168 @@ function _goBack(_ = Invalid as Dynamic) as Dynamic
...
+' If identifier is omitted, pops to the most recent route with any checkpoint.
+' All routes above the target are destroyed (including keepAlive views — no suspension).
+' No canDeactivate guards run.
+' @param {Dynamic} identifier - A string checkpoint identifier to search for, or Invalid to find any checkpoint
+' @returns {Dynamic} - A promise that resolves when navigation completes, or rejects if no checkpoint found
+function _popToCheckPoint(identifier = Invalid as Dynamic) as Dynamic
+    if m.__router_navigationInProgress then return _rejectNavigationInProgress()
+    m.__router_processingGoBack = true
+
+    ' Validate identifier type if provided
```
Comment (TwitchBronBron) [inline src/components/Router.bs]: Same thing as above, handle the `identifier = "__INVALID__"` feature to differentiate between omitted and actual `invalid`

File: src/components/Router.bs [hunk 6]
```diff
@@ -719,19 +868,23 @@ function showView(view as Object, onResume = false as Boolean) as Dynamic
...
 
-                if context.view.route.navigationState.fromPushState then
-                    m.__router_historyStack.push(context.view.route.path)
+                    if context.view.route.navigationState.fromPushState then
+                        m.__router_historyStack.push({ path: context.view.route.path, nodeId: context.view.id })
+                    end if
                 end if
             end if
 
             m.__router_processingGoBack = false
```
Comment (TwitchBronBron) [inline src/components/Router.bs:885]: I'm wondering if we don't actually need this `processingPopTo` setting. Any navigation backwards should set the .fromPopState flag, and we can likely just reuse the `__router_processingGoBack` flag (or rename that to be more generic like `isProcessingSomething` lol)

File: src/components/Router.bs [hunk 7]
```diff
@@ -139,56 +156,168 @@ function _goBack(_ = Invalid as Dynamic) as Dynamic
...
+    end if
+
+    ' Dispatch NavigationStart
+    dispatchRouterState(RouterState.NavigationStart, {
+        id: targetView.route.id
+        route: targetView.route
+    })
+
+    ' Truncate history stack to just up to and including the target
+    newStack = []
```
Comment (TwitchBronBron) [inline src/components/Router.bs]: Some of this might be redundant. Evaluate. :)

File: src/components/Router.bs [hunk 8]
```diff
@@ -139,56 +156,168 @@ function _goBack(_ = Invalid as Dynamic) as Dynamic
...
+    end function).catch(function(error as Dynamic) as Dynamic
+        m.__router_processingGoBack = false
+        m.__router_processingPopTo = false
+        m.__router_navigationInProgress = false
+        return promises.reject(error)
+    end function).toPromise()
+end function
+
 ' Function to navigate to a specific route
 ' @param {String|Object} path - The path to navigate to, or an AA for named routes: { name: String, params: Object }
```
Comment (TwitchBronBron) [inline src/components/Router.bs:310]: this should probably be assigned at the start of this function?

File: src/components/Router.bs [hunk 9]
```diff
@@ -139,56 +156,168 @@ function _goBack(_ = Invalid as Dynamic) as Dynamic
...
 ' @param {Object} options - The route config overrides and navigation options
 ' @returns {Dynamic} - The result of the navigation
```
Comment (TwitchBronBron) [inline src/components/Router.bs:312]: Can this be removed? (figure the router_processingGoBack assingment area could handle it?)

File: src/source/router.bs [hunk 10]
```diff
@@ -61,6 +61,36 @@ namespace sgRouter
...
     end function
 
+    ' Marks the current active route as a checkpoint, optionally with a string identifier.
+    ' If identifier is Invalid or omitted, uses the active route's path as the identifier.
+    ' A single route may register multiple identifiers; duplicates are ignored.
+    ' @since 0.2.0
+    ' @param identifier - A string identifier for the checkpoint, or Invalid to use the active route's path
+    ' @param options - The options to pass to the router
+    ' @param options.router - The router to use, if not provided, it will try to find the router in the current chain
+    sub setCheckPoint(identifier = Invalid as Dynamic, options = {} as Object)
```
Comment (TwitchBronBron) [inline src/source/router.bs]: Will need to take that `__INVALID__` string

File: src/source/router.bs [hunk 11]
```diff
@@ -61,6 +61,36 @@ namespace sgRouter
...
+
+    ' Unwinds the history stack to the most recent checkpointed route.
+    ' If identifier is provided, pops to the most recent route with that checkpoint identifier.
+    ' If identifier is omitted, pops to the most recent route with any checkpoint.
+    ' @since 0.2.0
+    ' @param identifier - The checkpoint identifier to search for, or Invalid to find any checkpoint
+    ' @param options - The options to pass to the router
+    ' @param options.router - The router to use, if not provided, it will try to find the router in the current chain
+    ' @return - A promise that resolves when navigation completes, or rejects if no matching checkpoint is found
+    function popToCheckPoint(identifier = Invalid as Dynamic, options = {} as Object) as Dynamic
```
Comment (TwitchBronBron) [inline src/source/router.bs]: `__INVALID__` thing here too

File: README.md [hunk 12]
```diff
@@ -437,6 +437,101 @@ The route snapshot is assembled by the router by parsing:
...
 
+---
+
+## ⏪ Checkpoints — Unwind the Stack to a Known Point
+
+Checkpoints let you mark a stack entry as a named "save point" and later unwind back to it in one call. This is useful for multi-step flows (e.g. checkout, onboarding, sign-in) where you want to jump back to a known entry point without calling `goBack` repeatedly.
+
+### 1) Mark a checkpoint
+
+Call `setCheckPoint` at any time to stamp the currently active stack entry:
```
Comment (TwitchBronBron) [inline README.md]: Let's change this to `setCheckpoint` (lower `p` since it's all one word)

**Review by TwitchBronBron — CHANGES_REQUESTED** [review]

File: src/components/Router.bs (hunk 9 above)
Comment (iObject) [reply src/components/Router.bs:312]: Not safely, no. Here's why the check still does real work:

path = "" (the default): isAA is false, so it skips named-route resolution and falls straight through to findMatchingRoute("", ...). That will happily try to match the empty/root path — potentially hitting a wildcard or 404 route — and then dispatch NavigationStart. That's an unintended navigation, not a clean rejection.

path = Invalid: same path — bypasses named-route resolution, hits findMatchingRoute with Invalid, which BrightScript coerces to "". Same problem.

The only case the check covers that's also caught downstream is an empty AA {} — resolveNamedRouteArg already rejects that with a WARN because path.name is Invalid. So that one is redundant, but it's not worth splitting.

The check is three lines and sits right at the entry point — I'd keep it where it is.

File: src/source/router.bs (hunk 10 above)
Comment (iObject) [reply src/source/router.bs]: done

File: src/source/router.bs (hunk 11 above)
Comment (iObject) [reply src/source/router.bs]: done

File: README.md (hunk 12 above)
Comment (iObject) [reply README.md]: Done

**Comment by iObject** [comment]
> ## Pull request overview
> Adds `popTo`/active-route support to the sgRouter library and wires it into a sample “signup + verify email” flow intended to redirect users to signup and then return them to the originally requested screen.
> 
> **Changes:**
> 
> * Add `sgRouter.popTo()` and `sgRouter.getActiveRoute()` APIs plus Router internals/tests to support popping back to a prior history entry (including named routes and keepAlive interactions).
> * Introduce `SignupScreen` and `VerifyEmailScreen`, and a `SignupGuard` intended to enforce email verification on protected routes.
> * Update sample app routes/buttons and extend `AuthManager` with an `isEmailVerified` field.
> 
> ### Reviewed changes
> Copilot reviewed 13 out of 13 changed files in this pull request and generated 6 comments.
> 
> Show a summary per file
> File	Description
> test-project/components/Screens/Welcome/WelcomeScreen.bs	Adjusts Welcome menu button labels.
> test-project/components/Screens/Signup/SignupScreen.xml	Adds Signup UI markup.
> test-project/components/Screens/Signup/SignupScreen.bs	Adds Signup navigation to verify step (passes redirect target via query param).
> test-project/components/Screens/Signup/VerifyEmailScreen.xml	Adds Verify Email UI markup.
> test-project/components/Screens/Signup/VerifyEmailScreen.bs	Sets `isEmailVerified` and attempts to `popTo()` the original destination.
> test-project/components/Scene/MainScene/MainScene.bs	Registers new signup routes and applies `SignupGuard` to details routes.
> test-project/components/Managers/SignupGuard/SignupGuard.xml	Declares the guard component interface.
> test-project/components/Managers/SignupGuard/SignupGuard.bs	Implements email-verification gate via `canActivate`.
> test-project/components/Managers/Auth/AuthManager.xml	Adds `isEmailVerified` field to AuthManager.
> src/source/router.bs	Exposes `popTo()` and `getActiveRoute()` on the public `sgRouter` namespace.
> src/components/Router.xml	Adds `_popTo` and `_getActiveRoute` to Router component interface.
> src/components/Router.bs	Implements `_popTo`, adds active-route getter, and adjusts history updates during popTo.
> src/source/router.spec.bs	Adds a test suite covering `popTo` behaviors (errors, keepAlive, named routes, hash handling).
> 💡 [Add Copilot custom instructions](/rokucommunity/sgRouter/new/master?filename=.github/instructions/*.instructions.md) for smarter, more guided reviews. [Learn how to get started](https://docs.github.com/en/copilot/customizing-copilot/adding-repository-custom-instructions-for-github-copilot).

@copilot update this with the recent changes

File: test-project/components/Managers/SignupGuard/SignupGuard.bs [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+
+' A canActivate guard that redirects users whose email is not verified to /signup,
+' cancelling the attempted navigation. The current path is passed as
+' redirectToOnVerify so VerifyEmailScreen can navigate back to it after verification.
+function canActivate(route = {} as Object) as Dynamic
+	if m.global.AuthManager.isEmailVerified then
+		return true
+	end if
+
+	sgRouter.setCheckpoint()
```
[the remark this answers is not in the document]
Comment (iObject) [reply test-project/components/Managers/SignupGuard/SignupGuard.bs:11]: setCheckpoint() will registered the current route path as the checkpoint

File: src/components/Router.bs (hunk 6 above)
Comment (iObject) [reply src/components/Router.bs:885]: removed

**Review by TwitchBronBron — APPROVED** [review]
